### PR TITLE
Bi-directional dshot on iomcu F103

### PR DESCRIPTION
**Pull request by andyp1per**
This PR adds support for bdshot telemetry on iomcus using the F103. There is one limitation - only channels 1-4,7-8 support telemetry. It is not possible to support 5-6 as these share a DMA channel with the fast UART interconnect.

Support is built into the iomcu dshot variant so to use you simply need to set ```BRD_IO_DSHOT``` and supply the appropriate ```SERVO_BLH_BDMASK```. ESC telemetry indexes are modified to reflect the fact that ESC's 1-4 and 7-8 can now provide telemetry.

EDT is working, so you can get temperature, current and voltage information from the ESCs using bdshot as well. To use set ```SERVO_DSHOT_ESC``` in the normal way.

This PR also fixes a few bugs found in the iomcu dshot support:

- mode masks were being set incorrectly
- mode masks were being updated too frequently
- the minimum bdshot cancel timeout was being set incorrectly to micros rather than ticks
- EDT was not working on AUX channels of boards with iomcu
- DShot LEDs were not working with EDT
- servo reversing and 3d was not working on FMU channels of boards with IOMCU

This work was generously sponsored by Holybro and builds on the existing dshot work kindly sponsored by CubePilot.

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 1]
```diff
@@ -295,6 +295,27 @@ __RAMFUNC__ void RCOutput::dshot_update_tick(virtual_timer_t* vt, void* p)
     chSysUnlockFromISR();
 }
 
+// calculate how much time remains in the current cycle
+sysinterval_t RCOutput::calc_ticks_remaining(pwm_group &group, uint64_t time_out_us, uint64_t cycle_length_us)
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:299]: cycle_length_us should be uint32_t

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 2]
```diff
@@ -1460,28 +1466,65 @@ void RCOutput::dshot_send_trampoline(void *p)
  */
 void RCOutput::rcout_thread() {
     // don't start outputting until fully configured
-    while (!hal.scheduler->is_system_initialized()) {
+    while (!hal.scheduler->is_system_initialized() || _dshot_period_us == 0) {
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:1469]: this is very odd

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 3]
```diff
@@ -1537,45 +1591,49 @@ __RAMFUNC__ void RCOutput::dshot_send_next_group(void* p)
...
     }
+    chSysUnlock();
 }
 
 /*
   deallocate DMA channel
  */
 void RCOutput::dma_deallocate(Shared_DMA *ctx)
 {
+    chSysLock();
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:1620]: do we need to broaden the lock time here?

File: libraries/AP_HAL_ChibiOS/RCOutput.h [hunk 4]
```diff
@@ -644,12 +666,15 @@ class ChibiOS::RCOutput : public AP_HAL::RCOutput
     uint16_t create_dshot_packet(const uint16_t value, bool telem_request, bool bidir_telem);
     void fill_DMA_buffer_dshot(dmar_uint_t *buffer, uint8_t stride, uint16_t packet, uint16_t clockmul);
 
+    // event to allow dshot cascading
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/RCOutput.h:669]: ifdef so DSHOT_CASCADE not exist or is zero on non-IOMCU dshot

File: libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp [hunk 5]
```diff
@@ -181,24 +204,159 @@ void RCOutput::bdshot_ic_dma_allocate(Shared_DMA *ctx)
...
     }
+    chSysUnlock();
 }
 
 /*
   deallocate DMA channel
  */
 void RCOutput::bdshot_ic_dma_deallocate(Shared_DMA *ctx)
 {
+    chSysLock();
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp:215]: narrower locking preferred

**Review by tridge — CHANGES_REQUESTED** [review]
not complete review ... complex stuff!

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp (hunk 1 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp]: Fixed

File: libraries/AP_HAL_ChibiOS/RCOutput.h (hunk 4 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.h:670]: Fixed

File: libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp (hunk 5 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp]: Fixed

File: libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp [hunk 6]
```diff
@@ -172,12 +194,12 @@ void RCOutput::bdshot_ic_dma_allocate(Shared_DMA *ctx)
             if (group.bdshot.ic_dma_handle[icuch] == ctx && group.bdshot.ic_dma[icuch] == nullptr) {
                 chSysLock();
                 group.bdshot.ic_dma[icuch] = dmaStreamAllocI(group.dma_ch[icuch].stream_id, 10, bdshot_dma_ic_irq_callback, &group);
-                chSysUnlock();
 #if STM32_DMA_SUPPORTS_DMAMUX
                 if (group.bdshot.ic_dma[icuch]) {
                     dmaSetRequestSource(group.bdshot.ic_dma[icuch], group.dma_ch[icuch].channel);
                 }
 #endif
+                chSysUnlock();
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp:202]: is this change needed?

File: libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp [hunk 7]
```diff
@@ -214,8 +236,10 @@ void RCOutput::bdshot_prepare_for_next_pulse(pwm_group& group)
             // no locking required
             group.bdshot.enabled = true;
         } else {
-            osalDbgAssert(!group.bdshot.ic_dma_handle[group.bdshot.curr_telem_chan]->is_locked(), "DMA handle is already locked");
-            group.bdshot.ic_dma_handle[group.bdshot.curr_telem_chan]->lock();
+            osalDbgAssert(!group.bdshot.curr_ic_dma_handle, "IC DMA handle has not been released");
+            group.bdshot.curr_ic_dma_handle = group.bdshot.ic_dma_handle[group.bdshot.curr_telem_chan];
+            osalDbgAssert(group.shared_up_dma || !group.bdshot.curr_ic_dma_handle->is_locked(), "IC DMA handle is already locked");
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp:241]: could shared_up_dma be removed when not on iomcu dshot?

File: libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp [hunk 8]
```diff
@@ -458,9 +504,17 @@ __RAMFUNC__ void RCOutput::bdshot_finish_dshot_gcr_transaction(virtual_timer_t*
...
-    // rotate to the next input channel
     group->bdshot.prev_telem_chan = group->bdshot.curr_telem_chan;
+    // rotate to the next input channel, we have to rotate even on failure otherwise
+    // we will not get data from active channels
     group->bdshot.curr_telem_chan = bdshot_find_next_ic_channel(*group);
+
+    if (group->bdshot.dma_tx_size > 0) {
+        group->dshot_state = DshotState::RECV_COMPLETE;
+    } else {
+        group->dshot_state = DshotState::RECV_FAILED;
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp:515]: comment worthwhile on why this is needed - when no EDT reply?

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp (hunk 6 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp:202]: All the others are done from I context, so I believe this is a correct change

File: libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp (hunk 7 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp:242]: Fixed

File: libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp [hunk 9]
```diff
@@ -424,6 +461,7 @@ void RCOutput::bdshot_config_icu_dshot(stm32_tim_t* TIMx, uint8_t chan, uint8_t
         break;
     }
 }
+#endif
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp]: Fixed

File: libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp (hunk 8 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp:521]: Fixed

File: libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp [hunk 10]
```diff
@@ -665,8 +733,16 @@ bool RCOutput::bdshot_decode_telemetry_from_erpm(uint16_t encodederpm, uint8_t c
...
     // eRPM = m << e (see https://github.com/bird-sanctuary/extended-dshot-telemetry)
-    uint8_t expo = ((encodederpm & 0xfffffe00U) >> 9U) & 0xffU;
-    uint16_t value = (encodederpm & 0x000001ffU);
+    uint8_t expo = ((encodederpm & 0xfffffe00U) >> 9U) & 0xffU; // 3bits
+    uint16_t value = (encodederpm & 0x000001ffU);               // 9bits
+#if HAL_WITH_ESC_TELEM
+    uint8_t normalized_chan = chan;
+#endif
+#if HAL_WITH_IO_MCU
+    if (AP_BoardConfig::io_dshot()) {
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp]: Fixed

File: libraries/AP_IOMCU/iofirmware/ioprotocol.h [hunk 11]
```diff
@@ -184,6 +189,13 @@ struct __attribute__((packed, aligned(2))) page_GPIO {
     uint8_t output_mask;
 };
 
+struct __attribute__((packed, aligned(2))) page_mode_out {
```
Comment (tridge) [inline libraries/AP_IOMCU/iofirmware/ioprotocol.h:192]: i think we should remove the packed, aligned attributes

File: libraries/AP_IOMCU/iofirmware/ioprotocol.h [hunk 12]
```diff
@@ -192,3 +204,17 @@ struct __attribute__((packed, aligned(2))) page_dshot {
     uint8_t repeat_count;
     uint8_t priority;
 };
+
+struct __attribute__((packed, aligned(2))) page_dshot_erpm {
```
Comment (tridge) [inline libraries/AP_IOMCU/iofirmware/ioprotocol.h:208]: doesn't need packed/aligned?

File: libraries/AP_IOMCU/iofirmware/ioprotocol.h [hunk 13]
```diff
@@ -184,6 +189,13 @@ struct __attribute__((packed, aligned(2))) page_GPIO {
...
 };
 
+struct __attribute__((packed, aligned(2))) page_mode_out {
+    uint16_t mask;
+    uint16_t mode;
+    uint16_t bdmask;
+    uint16_t esc_type;
+};
+
 struct __attribute__((packed, aligned(2))) page_dshot {
```
Comment (tridge) [inline libraries/AP_IOMCU/iofirmware/ioprotocol.h:199]: doesn't need packed/aligned

**Review by tridge — CHANGES_REQUESTED** [review]
mostly needs testing now

File: libraries/AP_IOMCU/iofirmware/ioprotocol.h (hunk 11 above)
Comment (andyp1per) [reply libraries/AP_IOMCU/iofirmware/ioprotocol.h]: Fixed

File: libraries/AP_IOMCU/iofirmware/ioprotocol.h (hunk 12 above)
Comment (andyp1per) [reply libraries/AP_IOMCU/iofirmware/ioprotocol.h]: Fixed

File: libraries/AP_IOMCU/iofirmware/ioprotocol.h (hunk 13 above)
Comment (andyp1per) [reply libraries/AP_IOMCU/iofirmware/ioprotocol.h]: Fixed

File: libraries/AP_IOMCU/AP_IOMCU.cpp [hunk 14]
```diff
@@ -371,6 +393,66 @@ void AP_IOMCU::read_rc_input()
...
+    }
+    uint8_t motor_poles = 14;
+    AP_BLHeli* blh = AP_BLHeli::get_singleton();
+    if (blh) {
+        motor_poles = blh->get_motor_poles();
+    }
+    for (uint8_t i = 0; i < IOMCU_MAX_CHANNELS/4; i++) {
+        for (uint8_t j = 0; j < 4; j++) {
+            const uint8_t esc_id = (i * 4 + j);
+            if (dshot_erpm.update_mask & 1U<<esc_id) {
```
Comment (tridge) [inline libraries/AP_IOMCU/AP_IOMCU.cpp:413]: should update_mask be cleared at the end of this loop? It looks like it is only cleared on timeout, which means it looks like we can call update_rpm multiple times per piece of data from the IOCMU

File: libraries/AP_IOMCU/AP_IOMCU.cpp [hunk 15]
```diff
@@ -371,6 +393,66 @@ void AP_IOMCU::read_rc_input()
...
+    }
+    for (uint i = 0; i<4; i++) {
+        TelemetryData t {
+            .temperature_cdeg = int16_t(telem->temperature_cdeg[i]),
+            .voltage = float(telem->voltage_cvolts[i]) * 0.01,
+            .current = float(telem->current_camps[i]) * 0.01
+        };
+        update_telem_data(esc_group * 4 + i, t, telem->types[i]);
+    }
+    esc_group = (esc_group + 1) % 4;
```
Comment (tridge) [inline libraries/AP_IOMCU/AP_IOMCU.cpp:453]: do we actually support telem from both esc groups? are we wasting half the bandwidth here by requesting telem from a group that can never give it?

File: libraries/AP_IOMCU/iofirmware/ioprotocol.h [hunk 16]
```diff
@@ -58,6 +58,11 @@ enum iopage {
     PAGE_MIXING = 200,
     PAGE_GPIO = 201,
     PAGE_DSHOT = 202,
+    PAGE_RAW_DSHOT_ERPM = 203,
+    PAGE_RAW_DSHOT_TELEM_1_4 = 204,
+    PAGE_RAW_DSHOT_TELEM_5_8 = 205,
+    PAGE_RAW_DSHOT_TELEM_9_12 = 206,
+    PAGE_RAW_DSHOT_TELEM_13_16 = 207,
```
Comment (tridge) [inline libraries/AP_IOMCU/iofirmware/ioprotocol.h:65]: why do we have pages for non-existant IOMCU groups?

**Review by tridge — CHANGES_REQUESTED** [review]
I'm seeing bad VServo in testing on Pixhawk6X-bdshot:
image
that is 27.3 volts

**Comment by tridge** [comment]
A few more issues:
saw this while testing with 6 dshot ESCs on pixhawk6x:
image
2nd problem is that we get this:
```
AP: RCOut: DS150:1-14 PWM:15-16
```
but motortest makes it clear that 2 of the motors are actually using PWM
ahh, if I refetch params after boot I get this:
```
AP: RCOut: DS150:1-4 PWM:5-8 DS150:9-14 PWM:15-16
```
so it changes after boot
full boot log:

**Comment by tridge** [comment]
image

**Comment by tridge** [comment]
tested the hexa with master Pixhawk6X-bdshot fw, with 6 motors on IOMCU, all 6 get dshot working, as confirmed with motors speeds in motortest.
with this PR, only 4 of the 6 get dshot outputs, and none get bdshot RPM telemetry

File: libraries/AP_IOMCU/AP_IOMCU.cpp (hunk 14 above)
Comment (andyp1per) [reply libraries/AP_IOMCU/AP_IOMCU.cpp:413]: The first line of the function reads into dshot_erpm overwriting whatever was there before. So I don't think it needs to be called.

File: libraries/AP_IOMCU/AP_IOMCU.cpp (hunk 15 above)
Comment (andyp1per) [reply libraries/AP_IOMCU/AP_IOMCU.cpp:453]: I have added defines for the telem channels so that we only use what is defined

File: libraries/AP_IOMCU/iofirmware/ioprotocol.h (hunk 16 above)
Comment (andyp1per) [reply libraries/AP_IOMCU/iofirmware/ioprotocol.h:65]: It was because we were defining IOMCU_MAX_CHANNELS as 16, I have now changed this to 8 and added an RC channels define

**Comment by andyp1per** [comment]
Can you post your config?

The internal error is usually because you changed something that requires a reboot.

**Comment by andyp1per** [comment]
Where does VServo come from? I'm guessing this is because you are not getting data on occasion

**Comment by tridge** [comment]
> Where does VServo come from? I'm guessing this is because you are not getting data on occasion

Vservo comes from an ADC pin on the IOMCU. It measured the voltage on the servo rail (primarily for servos). In this case there was no BEC on the servo rail (as ESCs don't need it) so voltage should have been zero

**Comment by tridge** [comment]
> Ok so in theory should be entirely unaffected by this PR ...

right, but it isn't

**Comment by andyp1per** [comment]
> right, but it isn't

Yes hence the "in theory" :)

**Comment by tridge** [comment]
> Can you post your config?

config for Pixhawk6X (CUAV version) here:
http://uav.tridgell.net/tmp/hexa.parm

**Comment by andyp1per** [comment]
@tridge 

- VServo fixed
- Stay on dshot if bdshot not supported implemented
- Support bdshot on iomcu on non-bdshot fmu implemented
- Fixed bug in zero erpm handling

File: libraries/AP_IOMCU/iofirmware/analog.cpp [hunk 17]
```diff
@@ -26,63 +26,105 @@
...
-    /* wait for the conversion to complete */
-    uint32_t counter = 16;
-
-    while (!(ADC1->SR & ADC_SR_EOC)) {
-        if (--counter == 0) {
-            // ensure EOC is clear
-            ADC1->SR = 0;
-            return 0xffff;
-        }
+    if (ADC1->SR & ADC_SR_STRT) {
```
Comment (tridge) [inline libraries/AP_IOMCU/iofirmware/analog.cpp:80]: I think this will happen a lot, and may cause us to get unreliable vservo/rssi updates. We call adc_sample_vrssi() and adc_sample_vservo() in quick succession, so the ADC will be busy most of the time on the vservo update if rssi sampling is done.
I think we need an adc_sample_next() call or similar which toggles between the two.

**Review by tridge — CHANGES_REQUESTED** [review]
please add a comment on what vrssi and vservo testing has been done. We rely on both being at high enough rates to diagnose radio and servo issues. I suspect with the method in this PR at the moment that it is sporadic

File: libraries/AP_IOMCU/iofirmware/analog.cpp (hunk 17 above)
Comment (andyp1per) [reply libraries/AP_IOMCU/iofirmware/analog.cpp:80]: yeah good call, the API seemed a bit anomalous with the interrupt changes

Comment (andyp1per) [reply libraries/AP_IOMCU/iofirmware/analog.cpp:98]: Fixed

**Comment by tridge** [comment]
@andyp1per analogin changes look right, but we will need to do some testing to ensure it is getting Vservo and RSSI at a good rate, without and without SBUS out (which is oddly linked to RSSI in)
did you work out if your signal generator you use for IMUs is suitable?

**Comment by andyp1per** [comment]
VRSSI mapped to VSERVO, unplug and replug:
image

VSERVO, unplug and replug:
image

I verified that setting BRD_SBUS_OUT sets VRSSI to 0

**Comment by tridge** [comment]
@andyp1per testing the PR from today:
image
that is CubeOrange quadplane

**Comment by andyp1per** [comment]
> @andyp1per testing the PR from today: image that is CubeOrange quadplane

Thanks @tridge - can you tell me what you did?

**Comment by andyp1per** [comment]
@tridge I tried copter on CubeOrange and it worked fine, so really need to know what your settings are so t hat I can reproduce.

**Comment by tridge** [comment]
@andyp1per sorry, here are the details
I was on this branch https://github.com/tridge/ardupilot/tree/iomcu-errors which includes this PR and several other needs-testing PRs
tlog of the session is here:
http://uav.tridgell.net/tmp/iomcu-err.tlog
parameters: http://uav.tridgell.net/tmp/iomcu.parm
it was a CubeOrange quadplane not using DShot. I included this PR just as part of my general testing and wasn't expecting an issue.

**Comment by andyp1per** [comment]
> it was a CubeOrange quadplane not using DShot. I included this PR just as part of my general testing and wasn't expecting an issue.

@tridge I can't reproduce the issue with your changes but I do see quite a long pause at startup. I can imagine that the combination of this pause plus the IOMCU update that you would have experienced switching to the new branch would lead to a timeout and the internal error. I would be interested to know if you still got it on reboot or whether the error was gone. My suggestion is that we increase the update timeout. I will try your branch as well.

**Comment by andyp1per** [comment]
@tridge I was able to reproduce this only once with your branch - I am 99% certain that the cause is the long delay at startup. Do you have any idea what is causing this? One of the issues with the iomcu startup code is that any long delay will cause a permanent error. If I load copter with default parameters there is no delay, so my guess is that it is some subsystem in conjunction with your parameter set.

**Comment by tridge** [comment]
@andyp1per most likely the CAN delay of 5s. We could call the IOMCU update fn in that delay loop

**Comment by tridge** [comment]
@andyp1per this code:
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_DroneCAN/AP_DroneCAN.cpp#L360

**Comment by andyp1per** [comment]
@tridge with my latest commit I can no longer reproduce the internal error

**Comment by juvinski** [comment]
Hi guys,

I had an strange error - and I almost lost my plane on last wednesday - without using dshot -
https://github.com/ArduPilot/ardupilot/issues/25548
Is this related ?

**Comment by andyp1per** [comment]
> Hi guys,
> 
> I had an strange error - and I almost lost my plane on last wednesday - without using dshot - #25548 Is this related ?

Unrelated - you were on 4.4

**Comment by andyp1per** [comment]
So the internal error appears to be some kind of race on the uart. It only occurs on the second packet of a multi-packet write_registers() call (in this case its the second packet of a MIXING frame). What I think happens is that as soon as the response is received for the first packet the second packet is sent out and the second packet response gets lost. On dshot we run the loop at 1Khz and only update faster than this if we receive an inbound event - the error doesn't occur under these circumstances. On non-dshot we immediately re-rerun the loop whilst a packet is already incoming and I think this is where the race is. Making it work like dshot makes the problem go away and seems like the right thing to do anyway. I think my other fixes, although correct, were simply introducing a bit more delay in sending out the second packet that also makes the problem go away. Note too that even in the case of an internal error the system recovers quickly. I have been unable to reproduce the issue with these fixes in place.

**Comment by timtuxworth** [comment]
I built master (not including this PR_ on my Durandal and got a number of internal errors and iomcu errors. If I use this PR, these errors go away. I don't know what that means, but I think no errors is good - right? 
Screenshot 2023-12-11 at 6 29 12 PM

**Review by tridge — APPROVED** [review]

**Comment by andyp1per** [comment]
This was tested by @timtuxworth at the weekend on plane on a Durandal with no issues found - thanks Tim
